File: bst/bst.cpp
#include <iostream>
#include <sstream>
#include <fstream>

#include <string>

#include <memory>

using namespace std;

struct bstnode {
    string word;
    int level;
    shared_ptr<bstnode> right = nullptr;
    shared_ptr<bstnode> left = nullptr;

    explicit bstnode(string w, int l) : word(move(w)), level(l){}
};

using nodeptr = shared_ptr<bstnode>;

bool chklevel;
ostringstream oss;

void insert(nodeptr* root, const string& word, int level)
{
    if(*root == nullptr){ // 트리가 비었을 때
        nodeptr newOne = make_shared<bstnode>(word, level);
        *root = newOne;
    }
    else if((*root)->word == word){ // 트리안의 문자열과 입력된 문자열이 같으면 덮어씀.
        (*root)->word = word;
    }
    else if((*root)->word > word){ // 트리안의 문자열보다 입력된 문자열이 작으면
        insert(&((*root)->left), word, ++level); // 왼쪽으로
    }

    else if((*root)->word < word){ // 트리안의 문자열보다 입력된 문자열이 크면
        insert(&((*root)->right), word, ++level); // 오른쪽으로
    }

}

nodeptr* findNode(const nodeptr& root, nodeptr* sub)
{
    nodeptr tNode;
    if(root->left!=nullptr){ // 왼쪽 부트리가 있으면
        tNode = root->left;
        sub = &(root->left);
        while(tNode->right != nullptr){ // 그 부트리 중에서 최댓값을 찾음
            sub = &(tNode->right);
            tNode = tNode->right;
        }
    }
    else if(root->right != nullptr){ // 오른쪽 부트리가 있으면
        tNode = root->right;
        sub = &(root->right);
        while(tNode->left != nullptr){  // 그 부트리 중에서 최솟값을 찾음
            sub = &(tNode->left);
            tNode = tNode->left;
        }
    }
    return sub;
}

void remove(nodeptr* root){
    if((*root)->left == nullptr && (*root)->right == nullptr)
        *root = nullptr; // 양쪽 부트리가 모두 없으면(leaf라면) 루트를 없앰
    else { // leaf까지 찾은 다음 없앰
        nodeptr* tNode = nullptr;
        tNode = findNode((*root), tNode);
        (*root)->word = (*tNode)->word;
        remove(tNode);
    }
}

void delNode(nodeptr* root, const string& word)
{
    if(*root != nullptr){
        if((*root)->word == word) // 입력된 문자열과 일치하면 없앰
            remove(root);
        else if((*root)->word < word) // 입력된 문자열보다 작으면
            delNode(&((*root)->right), word); // 왼쪽으로
        else if((*root)->word > word) // 입력된 문자열보다 크면
            delNode(&((*root)->left), word); // 오른쪽으로
    }
}

void leaf(nodeptr* root)
{
    if((*root)->left==nullptr && (*root)->right==nullptr)
        oss << (*root)->word << " ";
    if((*root)->left!=nullptr)
        leaf(&((*root)->left));
    if((*root)->right!=nullptr)
        leaf(&((*root)->right));
}

void findDepth(nodeptr* root, int lev)
{
    if((*root)->level == lev){
        oss << (*root)->word << " ";
        chklevel = true;
    }
    if((*root)->left!=nullptr)
        findDepth(&((*root)->left), lev);
    if((*root)->right!=nullptr)
        findDepth(&((*root)->right), lev);
}

int main()
{
    ifstream ifs("bst.inp", ios::binary);
    ofstream ofs("bst.out", ios::binary);
    if(!ifs.is_open()){
        ofs << "missing input file...";
        ofs.close();
        ifs.close();
        return EXIT_FAILURE;
    }
    nodeptr root = nullptr;
    int count=0, kdepth=0;
    int deflev=1; // default node level
    string cmd, word, result;
    ifs >> count;
    while(count > 0 && ifs >> cmd){
        if(cmd == "+"){
            ifs >> word;
            insert(&root, word, deflev);
        }
        else if(cmd == "-"){
            ifs >> word;
            delNode(&root, word);
        }
        else if(cmd == "depth"){
            ifs >> kdepth;
            chklevel = false;
            findDepth(&root, kdepth);
            if(!chklevel){
                result += "NO\n";
            }
            else{
                result += oss.str();
                result.pop_back();
                result+='\n';
            }
            oss.str("");
            oss.clear();
        }
        else if(cmd == "leaf"){
            leaf(&root);
            result += oss.str();
            result.pop_back();
            result+='\n';
            oss.str("");
            oss.clear();
        }
    }
    result.pop_back();
    ofs << result;
    ofs.close();
    ifs.close();
    return EXIT_SUCCESS;
}

File: bst/bst.py
#-*-coding:utf-8-*-

class Node:
    def __init__(self, name, depth):
        self.name = name
        self.depth = depth
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None
        self.leaflist = []
        self.depmap = {}

    def insert(self, inpdata, depth):
        self.root = self.insnode(self.root, inpdata, depth)
        return self.root is not None

    def insnode(self, node, inpdata, level):
        if node is None:
            node = Node(inpdata, level)
        else :
            if node.name <= inpdata :
                level+=1
                node.left = self.insnode(node.left, inpdata, level)
            else :
                level+=1
                node.right = self.insnode(node.right, inpdata, level)
        return node

    def findleaf(self):
        self.getleaf(self.root)
        return " ".join(self.leaflist)

    def getleaf(self, root):
        if root.left is None and root.right is None :
            self.leaflist.append(root.name)
        if root.left is not None :
            root = self.getleaf(root.left)
        if root.right is not None :
            root = self.getleaf(root.right)

    def depth(self, key):
        deps = []
        self.finddepth(self.root, key)
        i = len(self.depmap)
        while i > 0:
            if self.depmap[key] is None :
                deps.append("NO")
            else:
                deps.append(self.depmap[key])
            i-=1
        return " ".join(deps)

    def finddepth(self, root, k):
        if root.depth == k :
            self.depmap[k] = root.name
        if root.left is not None :
            self.finddepth(root.left, k)
        if root.right is not None :
            self.finddepth(root.right, k)

    def findMaxMin(self, root):
        temp = sub = None
        if root.left is not None :
            temp = root.left
            sub = root.left
            while temp.right is not None :
                sub = temp.right
                temp = temp.right
        elif root.right is not None :
            temp = root.right
            sub = root.right
            while temp.left is not None :
                sub = temp.left
                temp = temp.left
        return sub

    def remove(self, root):
        if root.left is None and root.right is None :
            root = None
        else :
            sub = self.findMaxMin(root)
            root.name = sub.name
            self.remove(root)

    def delNode(self, root, inpdata):
        if root is not None:
            if root.name == inpdata :
                self.remove(root)
            elif root.name > inpdata :
                self.delNode(root.left, inpdata)
            elif self.root.name < inpdata :
                self.delNode(root.right, inpdata)

def main():
    cmdList = []
    with open("bst.inp", "rb") as inp :
        cmdNum = int(inp.readline())
        for line in inp :
            cmdList.append(line.rstrip())
    defDepth = 1

    bst = BinarySearchTree()
    result = ""

    for line in cmdList :
        if "+" in line:
            cutCmd = line.split()
            cutCmd.pop(0) # delete "+"
            key = cutCmd[0]
            bst.insert(key, defDepth)
        elif "-" in line:
            cutCmd = line.split()
            cutCmd.pop(0)  # delete "-"
            key = cutCmd[0]
            bst.delNode(bst.root,key)
        elif "depth" in line:
            cutCmd = line.split()
            cutCmd.pop(0)
            key = cutCmd[0]
            result+=bst.depth(key)+"\n"
        elif "leaf" in line:
            result+=bst.findleaf()+"\n"

    result.rstrip()
    with open("bst.out", "wb") as out:
        out.write(result)
    return

if __name__ == '__main__':
    main()